### ISSUE TITLE
realtime_interval and key_repeat_interval are browser-dependent

### ISSUE THREAD

**Issue by pancelor** [increpare/PuzzleScript#949]
A comment from a [puzzlescript dev](https://le-slo.itch.io/not-that-common-egg-eater-snake) that I just stumbled across on discord:

> It's weird that in firefox it's like 3x times slower, hard to nail the gamefeel for all browsers alike. Sorry to those who played on Chrome!

Apparently `key_repeat_interval ` acts differently across browsers? So I made a test file: [key_repeat_interval.txt](https://github.com/increpare/PuzzleScript/files/10269063/key_repeat_interval.txt). I set up a situation with `key_repeat_interval` that should have taken 10 seconds, and manually timed it (with my phone's stopwatch) in each browser. In firefox it took 10.03 seconds, but in chrome it took 13.21 seconds. that's much much slower than it should be!

(I observed similar results for [realtime_interval.txt](https://github.com/increpare/PuzzleScript/files/10269064/realtime_interval.txt) -- 10.00 seconds in firefox versus 14.17 seconds in chrome. I did not test `again_interval` but I would guess it has similar issues)

---

I suspect this happens because puzzlescript uses `setTimeout(loop,17)` to draw frames; as far as I know `requestAnimationFrame` is the "correct" way to run a HTML canvas game's mainloop. I don't know where I first heard that but I did a bit of searching and found some web resources that seem to agree:
- [this tutorial](https://spicyyoghurt.com/tutorials/html5-javascript-game-development/create-a-proper-game-loop-with-requestanimationframe) talks a little bit about why you would want `requestAnimationFrame` instead of `setInterval`
- [MainLoop.js](https://github.com/IceCreamYou/MainLoop.js), a game loop library that looks well-thought-out, and uses `requestAnimationFrame`

**Comment by pancelor** [comment]
this timing discrepancy seems like strong enough motivation to revive #924 / #925, so I did that: #948 is the new PR. I tested #948 locally and the discrepancy was fixed:

```
## results from key_repeat_interval.pz, after #948
firefox: 10.16 seconds
chrome:  10.13 seconds (yay)

## results from realtime_interval.pz, after #948
firefox: 10.13 seconds
chrome:  10.08 seconds (yay)
```

sorry to spread this conversation across 4 different places! I'll leave this thread here, focused on the bug, and I'll talk about requestAnimationFrame more as a reply to #925

**Comment by increpare** [comment]
I accepted the pull request - here are my timings + protocol for posterity:

my protocol - I take this file https://www.puzzlescript.net/play.html?p=155ad5f19dd7b1979cdee3b7fde6f9a9

I have a MacBook Pro M1 with 15.1 (24B83) and too much ram.

I do a screen recording of myself holding right, and I measure the time between when I first hit the goal on the left and when I reach the one on the right (using quicktime crop interface - it's a bit inexact because it's a mouse interface rather than a per-frame thing, but it's reasonably exact)

CURRENT VERSION ON PUZZLESCRIPT.NET (all tests on macos):

safari Version 18.1 (20619.2.8.11.10):
run 1: 1.93
run 2: 1.94
run 3: 1.91
average: 1.93

chrome Version 131.0.6778.86 (Offizieller Build) (arm64):
run 1: 1.96
run 2: 1.96
run 3: 1.96 
average: 1.96 

firefox 133.0 (aarch64):
run 1: 2.19
run 2: 2.20
run 3: 2.18
average: 2.19

NEW VERSION WITH YOUR CHANGES (all tests on macos):

safari Version 18.1 (20619.2.8.11.10):
run 1: 1.91
run 2: 1.89
run 3: 1.86
average: 1.89

chrome Version 131.0.6778.86 (Offizieller Build) (arm64):
run 1: 1.81
run 2: 1.83
run 3: 1.83
average: 1.82


firefox 133.0 (aarch64):
run 1: 1.83
run 2: 1.85
run 3: 1.84
average: 1.84

WHAT SHOULD WE GET with key_repeat_interval 0.2?  0.2*9 = 1.8 seconds.  So, pretty good! insofar as safari is a bit slow, it's less according to spec ^^

**Comment by increpare** [comment]
@pancelor thanks a lot for looking into this, reporting it, implementing a fix ,and rigorous testing <3

**Comment by increpare** [comment]
( The other thing I was worried about was the slight increase in general resource consumption caused by more frequent ticking, but I'd much prefer consistent input behaviour across platforms at the cost of slightly increased cpu use. )
